gateway/app/drivers/quality_on_demand/nef/nef_interface.py

```diff
@@ -0,0 +1,359 @@
+            qosReference=sub_info.qosProfile,
+            usageThreshold=UsageThreshold(
+                duration=sub_info.duration + req.requestedAdditionalDuration,
+                totalVolume=0,
+                downlinkVolume=0,
+                uplinkVolume=0,
+            ),
+        )
+
+        if sub_info.device.ipv4Address and sub_info.device.ipv4Address.publicAddress:
+            payload.ueIpv4Addr = sub_info.device.ipv4Address.publicAddress
+        elif sub_info.device.ipv6Address:
+            payload.ueIpv6Addr = sub_info.device.ipv6Address
+        elif sub_info.device.phoneNumber:
+            phone_number = sub_info.device.phoneNumber.lstrip("+")
+            payload.gpsi = f"msisdn-{phone_number}"
+
+        res = await self.httpx_client.put(
+            nef_id,
+            content=payload.model_dump_json(exclude_unset=True),
+            headers={"Content-Type": "application/json"},
+        )
+
+        if res.status_code == 404:
+            raise ResourceNotFound()
+
+        if not res.is_success:
+            raise BadRequest()
+
+        if sub_info.expiresAt is None:
+            raise ResourceNotFound()
+
+        response = SessionInfo(
+            sessionId=sub_info.sessionId,
+            device=sub_info.device,
+            applicationServer=sub_info.applicationServer,
+            devicePorts=sub_info.devicePorts,
+            applicationServerPorts=sub_info.applicationServerPorts,
+            qosProfile=sub_info.qosProfile,
+            duration=sub_info.duration + req.requestedAdditionalDuration,
+            startedAt=sub_info.startedAt,
+            expiresAt=sub_info.expiresAt
+            + datetime.timedelta(seconds=req.requestedAdditionalDuration),
+            qosStatus=QosStatus.REQUESTED,
+        )
+        await self._save_subscription_info(id, response)
+
+        return response
+
+    async def send_callback_message(
+        self, body: UserPlaneNotificationData, id: str
+    ) -> None:
+        sub_info = await self.get_qod_information_by_id(id)
+
+        if sub_info is None:
+            raise ResourceNotFound()
+
+        status = self._get_qod_status(body)
+        if status is not None:
+            sub_info.qosStatus = status
+
+        if sub_info.qosStatus == QosStatus.UNAVAILABLE:
+            sub_info.statusInfo = (
+                StatusInfo.NETWORK_TERMINATED
+                if sub_info.qosStatus == QosStatus.UNAVAILABLE
+                and sub_info.statusInfo != StatusInfo.DELETE_REQUESTED
+                else sub_info.statusInfo
+            )
+
+        cloud_event = CloudEvent(
+            id=str(uuid.uuid4()),
+            source=f"{settings.gateway_url}qod-provisioning/v0.2/device-qos/{id}",
+            type=Type.org_camaraproject_quality_on_demand_v1_qos_status_changed,
+            specversion=Specversion.field_1_0,
+            datacontenttype=Datacontenttype.application_json,
+            data=sub_info.model_dump(mode="json", exclude_unset=True),
```

Comment:
This is just wrong the data doesn't expect the subscription but a different object entirely.